### PR TITLE
ynh_mongo_db

### LINKED ISSUE #63: Improve mongodb helper
- [ ] using ynh_handle_getopts_args 
- [ ] implement ynh_mongo_execute_as_root
- [ ] implement ynh_mongo_execute_file_as_root
or
- [ ] implement ynh_mongo_restore_db

### PR DESCRIPTION
**Pull request by yalh76**
Improve #19 and Fix #63 

List of mysql/postgresql functions:
~~- [ ] ynh_mongo_connect_as~~
~~- [ ] ynh_mongo_execute_as_root~~
~~- [ ] ynh_mongo_execute_file_as_root~~
- [x] ynh_mongo_create_db
- [x] ynh_mongo_drop_db
- [x] ynh_mongo_dump_db
- [x] ynh_mongo_create_user
~~- [ ] ynh_mongo_user_exists~~
- [x] ynh_mongo_database_exists
- [x] ynh_mongo_drop_user
- [x] ynh_mongo_setup_db
- [x] ynh_mongo_remove_db
~~- [ ] ynh_mongo_test_if_first_run~~

New one:
- [x] ynh_mongo_restore_db

### CODE REVIEW

**Comment by yalh76** [comment]
`ynh_mongo_user_exists` is like impossible to implement, depending of the version of mongo, not all functions are available

File: ynh_mongo_db/ynh_mongo_db [hunk 1]
```diff
@@ -0,0 +1,270 @@
...
+}
+
+# Drop a database
+#
+# [internal]
+#
+# If you intend to drop the database *and* the associated user,
+# consider using ynh_mongo_remove_db instead.
+#
+# usage: ynh_mongo_drop_db -database=db
```
Comment (ericgaspar) [inline ynh_mongo_db/ynh_mongo_db]:
```suggestion
# usage: ynh_mongo_drop_db -database=db_name
```
Same terminology as line 75 & 130

File: ynh_mongo_db/ynh_mongo_db [hunk 2]
```diff
@@ -0,0 +1,270 @@
...
+    ynh_handle_getopts_args "$@"
+
+    ynh_mongo_exec_as --database="${database}" --command='db.runCommand({dropDatabase: 1})'
+}
+
+# Dump a database
+#
+# example: ynh_mongo_dump_db --database=wekan > ./dump.bson
+#
+# usage: ynh_mongo_dump_db --database=database
```
Comment (ericgaspar) [inline ynh_mongo_db/ynh_mongo_db:140]:
```suggestion
# usage: ynh_mongo_dump_db --database=db_name
```

File: ynh_mongo_db/ynh_mongo_db [hunk 3]
```diff
@@ -0,0 +1,270 @@
...
+    # Create the user and set the user as admin of the db
+    ynh_mongo_exec_as --database="$db_name" --command='db.createUser( { user: "'${db_user}'", pwd: "'${db_pwd}'", roles: [ { role: "readWrite", db: "'${db_name}'" } ] } );'
+    
+    # Add clustermonitoring rights
+    ynh_mongo_exec_as --database="$db_name" --command='db.grantRolesToUser("'${db_user}'",[{ role: "clusterMonitor", db: "admin" }]);'
+}
+
+# Check if a mongo database exists
+#
+# usage: ynh_mongo_database_exists --database=database
```
Comment (ericgaspar) [inline ynh_mongo_db/ynh_mongo_db:185]:
```suggestion
# usage: ynh_mongo_database_exists --database=db_name
```

File: ynh_mongo_db/ynh_mongo_db [hunk 4]
```diff
@@ -0,0 +1,270 @@
...
+    else
+        return 1
+    fi
+}
+
+# Restore a database
+#
+# example: ynh_mongo_restore_db --database=wekan < ./dump.bson
+#
+# usage: ynh_mongo_restore_db --database=database
```
Comment (ericgaspar) [inline ynh_mongo_db/ynh_mongo_db:210]:
```suggestion
# usage: ynh_mongo_restore_db --database=db_name
```

File: ynh_mongo_db/ynh_mongo_db [hunk 5]
```diff
@@ -0,0 +1,270 @@
...
+    local db_user
+    # Manage arguments with getopts
+    ynh_handle_getopts_args "$@"
+
+    ynh_mongo_exec_as --command='db.dropUser("'${db_user}'", {w: "majority", wtimeout: 5000})'
+}
+
+# Create a database, an user and its password. Then store the password in the app's config
+#
+# usage: ynh_mongo_setup_db --db_user=user --db_name=name [--db_pwd=pwd]
```
Comment (ericgaspar) [inline ynh_mongo_db/ynh_mongo_db:248]:
```suggestion
# usage: ynh_mongo_setup_db --db_user=db_user --db_name=db_name [--db_pwd=db_pwd]
```

File: ynh_mongo_db/ynh_mongo_db [hunk 6]
```diff
@@ -0,0 +1,270 @@
...
+    
+    # Create the user and grant access to the database
+    ynh_mongo_create_user --db_user="$db_user" --db_pwd="$db_pwd" --db_name="$db_name"
+
+    ynh_app_setting_set --app=$app --key=db_pwd --value=$db_pwd 
+}
+
+# Remove a database if it exists, and the associated user
+#
+# usage: ynh_mongo_remove_db --db_user=user --db_name=name
```
Comment (ericgaspar) [inline ynh_mongo_db/ynh_mongo_db:280]:
```suggestion
# usage: ynh_mongo_remove_db --db_user=db_user --db_name=db_name
```

File: ynh_mongo_db/ynh_mongo_db [hunk 7]
```diff
@@ -0,0 +1,270 @@
...
+    ynh_handle_getopts_args "$@"
+
+    mongorestore --quiet --db="$database" --archive
+}
+
+# Drop a user
+#
+# [internal]
+#
+# usage: ynh_mongo_drop_user --db_user=user
```
Comment (ericgaspar) [inline ynh_mongo_db/ynh_mongo_db]:
```suggestion
# usage: ynh_mongo_drop_user --db_user=db_user
```

File: ynh_mongo_db/ynh_mongo_db [hunk 8]
```diff
@@ -0,0 +1,270 @@
...
+# | arg: -u, --db_user=     -the user to drop
```
Comment (ericgaspar) [inline ynh_mongo_db/ynh_mongo_db]:
```suggestion
# | arg: -u, --db_user=     - The user to drop
```

**Review by ericgaspar — COMMENTED** [review]
In the usage sections, the database name is called: `db`, `db_name`, `database`, `name`.
I wonder if it would help to have a unified terminology...

File: ynh_mongo_db/ynh_mongo_db [hunk 9]
```diff
@@ -0,0 +1,270 @@
...
+        return 1
+    fi
+}
+
+# Restore a database
+#
+# example: ynh_mongo_restore_db --database=wekan < ./dump.bson
+#
+# usage: ynh_mongo_restore_db --database=database
+# | arg: -d, --database=    - the database name to restore
```
Comment (ericgaspar) [inline ynh_mongo_db/ynh_mongo_db]:
```suggestion
# | arg: -d, --database=    - The database name to restore
```

File: ynh_mongo_db/ynh_mongo_db [hunk 10]
```diff
@@ -0,0 +1,270 @@
...
+    ynh_mongo_exec_as --database="$db_name" --command='db.createUser( { user: "'${db_user}'", pwd: "'${db_pwd}'", roles: [ { role: "readWrite", db: "'${db_name}'" } ] } );'
+    
+    # Add clustermonitoring rights
+    ynh_mongo_exec_as --database="$db_name" --command='db.grantRolesToUser("'${db_user}'",[{ role: "clusterMonitor", db: "admin" }]);'
+}
+
+# Check if a mongo database exists
+#
+# usage: ynh_mongo_database_exists --database=database
+# | arg: -d, --database=    - the database for which to check existence
```
Comment (ericgaspar) [inline ynh_mongo_db/ynh_mongo_db]:
```suggestion
# | arg: -d, --database=    - The database for which to check existence
```

File: ynh_mongo_db/ynh_mongo_db [hunk 11]
```diff
@@ -0,0 +1,270 @@
...
+    mongodump --quiet --db="$database" --archive
+}
+
+# Create a user
+#
+# [internal]
+#
+# usage: ynh_mongo_create_user --db_user=db_user --db_pwd=db_pwd --db_name=db_name
+# | arg: -u, --db_user=    - the user name to create
+# | arg: -p, --db_pwd=     - the password to identify user by
```
Comment (ericgaspar) [inline ynh_mongo_db/ynh_mongo_db]:
```suggestion
# | arg: -u, --db_user=    - The user name to create
# | arg: -p, --db_pwd=     - The password to identify user by
```

File: ynh_mongo_db/ynh_mongo_db [hunk 12]
```diff
@@ -0,0 +1,270 @@
...
+
+    ynh_mongo_exec_as --database="${database}" --command='db.runCommand({dropDatabase: 1})'
+}
+
+# Dump a database
+#
+# example: ynh_mongo_dump_db --database=wekan > ./dump.bson
+#
+# usage: ynh_mongo_dump_db --database=database
+# | arg: -d, --database=    - the database name to dump
```
Comment (ericgaspar) [inline ynh_mongo_db/ynh_mongo_db]:
```suggestion
# | arg: -d, --database=    - The database name to dump
```

File: ynh_mongo_db/ynh_mongo_db [hunk 13]
```diff
@@ -0,0 +1,270 @@
...
+
+# Drop a database
+#
+# [internal]
+#
+# If you intend to drop the database *and* the associated user,
+# consider using ynh_mongo_remove_db instead.
+#
+# usage: ynh_mongo_drop_db -database=db
+# | arg: -d, --database=    - the database name to drop
```
Comment (ericgaspar) [inline ynh_mongo_db/ynh_mongo_db]:
```suggestion
# | arg: -d, --database=    - The database name to drop
```

File: ynh_mongo_db/ynh_mongo_db [hunk 14]
```diff
@@ -0,0 +1,270 @@
...
+quit()
+EOF
+}
+
+# Create a database 
+#
+# [internal]
+#
+# usage: ynh_mongo_create_db db_name
+# | arg: db_name - the database name to create
```
Comment (ericgaspar) [inline ynh_mongo_db/ynh_mongo_db]:
```suggestion
# | arg: db_name    - The database name to create
```

File: ynh_mongo_db/ynh_mongo_db [hunk 15]
```diff
@@ -0,0 +1,270 @@
...
+# example: ynh_mongo_exec_as --command="db.getMongo().getDBNames().indexOf(\"wekan\")"
+#
+# usage: ynh_mongo_exec_as [--user=user] [--password=password] [--authenticationdatabase=authenticationdatabase] [--database=database] [--host=host] [--port=port] --command="command"
+# | arg: -u, --user=                      - the user name to connect as
+# | arg: -p, --password=                  - the user password
+# | arg: -d, --authenticationdatabase=    - the authenticationdatabase to connect to
+# | arg: -d, --database=                  - the database to connect to
+# | arg: -h, --host=                      - the host to connect to
+# | arg: -P, --port=                      - the port to connect to
+# | arg: -c, --command=                   - the command to evaluate
```
Comment (ericgaspar) [inline ynh_mongo_db/ynh_mongo_db]:
```suggestion
# | arg: -u, --user=                      - The user name to connect as
# | arg: -p, --password=                  - The user password
# | arg: -d, --authenticationdatabase=    - The authenticationdatabase to connect to
# | arg: -d, --database=                  - The database to connect to
# | arg: -h, --host=                      - The host to connect to
# | arg: -P, --port=                      - The port to connect to
# | arg: -c, --command=                   - The command to evaluate
```

**Review by ericgaspar — COMMENTED** [review]
- *Typos*

**Pull request merged**